csharp/src/Ice/Properties.cs

```diff
@@ -126,19 +126,114 @@ internal PropertyValue(string v, bool u)
+            foreach (var entry in updates)
+            {
+                ValidateProperty(entry.Key, entry.Value);
+            }
+
+            lock (_properties)
+            {
+                foreach (var entry in updates)
+                {
+                    if (!SetPropertyImpl(entry.Key, entry.Value))
+                    {
+                        // Starting with .NET 3.0, Remove does not invalidate enumerators
+                        updates.Remove(entry.Key);
+                    }
+                }
+            }
+        }
+
+        /// <summary>Remove a property.</summary>
+        /// <param name="key">The property key.</param>
+        /// <returns>The property value or null if this property was not set.</returns>
+        public string? RemoveProperty(string key)
+        {
+            if (_properties.TryGetValue(key, out var pv))
+            {
+                _properties.Remove(key);
+                return pv.Val;
+            }
+            return null;
+        }
+
+        /// <summary>Get all properties that were not read.</summary>
+        /// <returns>The properties that were not read as a list of keys.</returns>
+        public List<string> GetUnusedProperties()
+        {
+            lock (_properties)
+            {
+                List<string> unused = new List<string>();
+                foreach (KeyValuePair<string, PropertyValue> entry in _properties)
+                {
+                    if (!entry.Value.Used)
+                    {
+                        unused.Add(entry.Key);
+                    }
+                }
+                return unused;
+            }
+        }
+
+        // SetPropertyImpl sets a property and returns true when the property
+        // was added, changed or removed, and false otherwise.
+        private bool SetPropertyImpl(string key, string value)
+        {
+            // Must be called with validated a validated property and with _properties locked
+
+            key = key.Trim();
+            Debug.Assert(key.Length > 0);
+            if (value.Length == 0)
+            {
+                return _properties.Remove(key);
+            }
+            else
+            {
+                // These properties are always marked "used"
+                bool used = (key == "Ice.ConfigFile" || key == "Ice.ProgramName");
+
+                if (_properties.TryGetValue(key, out var pv))
+                {
+                    if (!pv.Val.Equals(value))
+                    {
+                        pv.Val = value;
+                        pv.Used = used;
+                        return true;
+                    }
+                    // else Val == value, nothing to do
+                }
+                else
+                {
+                    _properties[key] = new PropertyValue(value, used);
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private void ValidateProperty(string key, string value)
```

Comment:
I think we should rename this method to `ValidatePropertyName` and don't pass the unused value